README.md

```diff
@@ -217,6 +217,41 @@ variant { Err = variant { BadBurn = record { min_burn_amount = ... } } }
 
 The minting account is also the receiver of the fees burnt in regular transfers.
 
+## Textual representation of accounts
+
+We specify a _canonical textual format_ that all applications should use to display ICRC-1 accounts.
+This format relies on the textual encoding of principals specified in the [Internet Computer Interface Specification](https://internetcomputer.org/docs/current/references/ic-interface-spec/#textual-ids), referred to as `Principal.toText` and `Principal.fromText` below.
+
+### Encoding
+
+Applications SHOULD encode accounts as follows:
+
+  1. The encoding of accounts that do not have a subaccount, such as `{ owner = principal "aaaaa-aa"; subaccount = null }`, is the encoding of the owner principal.
+  2. The encoding of accounts with a non-empty subaccount is the encoding of the concatenation of the owner principal bytes and the subaccount bytes.
+
+In pseudocode:
+
+```sml
+encodeAccount({ owner; subaccount = null      }) = Principal.toText(owner)
+encodeAccount({ owner; subaccount = opt bytes }) = Principal.toText(owner · bytes)
```

Comment:
I don't think this is a good idea. The trailing byte of a principal determines its type, and appending an arbitrary byte string to it will create confusion.

Reply:
Do you think we should add a new tag for such "extended" principals?
One property of that approach that we find very attractive is that each principal is a valid ICRC-1 account; I'd like to keep this property.

Reply:
@bogwar convinced me that there is no ambiguity here, since the length of the encoded string is not a valid principal – principals have a bounded length of 29 bytes and sub-accounts are 32 bytes.

I'm still not a big fan of this since I still think it may create confusion on the user side (e.g. a subaccount of the anonymous principal `0x04` is visually almost the same as a self-authenticating principal), but I also don't see a better option for now.

Reply:
First of all I find the sheer length of it very unattractive from a perspective of user experience. The use case where all 32 bytes of the subaccount id space are actually used is niche. With this solution we are forcing everyone to deal with encodings of this length. There has to be a better way. Users who only use subaccounts 1,2,3 don't have to be bothered with encodings of this length (and with, to them, inexplicably many a's).

I still vote for `4kydj-ryaaa-aaaag-qaf7a-cai:1`, etc., i.e. having the two components exposed. If principals really have to be valid account ids (I am not totally convinced why) then `4kydj-ryaaa-aaaag-qaf7a-cai` can be allowed as a shortcut for `4kydj-ryaaa-aaaag-qaf7a-cai:0`.

Reply:
@timohanke your proposals are attractive since they look much nicer. But:

* I would imagine that a common use is for users to make payments to services -- in this case they most likely would have to deal with long subaccounts (if services embrace the payment flow where each payment goes into a specific subaccount). 

* they lack a checksum over the entire string. Wouldn't that be an issue with copy pasting? (Granted, the issue is somewhat alleviated since the transfer would still go to the intended principal, even if to the wrong subaccout)

Reply:
So let's say there are two different use cases, one where the subaccount is small (including the case when it is 0 and you when would like to not have to mention it at all) and the other one where it is large (20 bytes+). Isn't it then also ok to have two different encoding schemes? The encoding doesn't have to be unique. The first scheme can use the explicit version where the two components are visible. And that one is probably fine without a checksum. The second scheme can use a long homogeneous looking string. If the use cases are sufficiently different then it would be ok to have different encodings. Then you also have more freedom in the second scheme. E.g. property 1 doesn't have to embed as nicely anymore into the second scheme because it already does into the first scheme.

Reply:
As for the original proposal, I noticed that in order to achieve property 1) (that existing principals are valid account ids) there is a switch between two cases happening: if the length is <= 29 bytes then it is interpreted as a principal and the subaccount 0 is assumed, if the length is >= 33 bytes then the last 32 bytes are taken as the subaccount and all bytes before that are taken as the principal. If such a switch is ok then we can achieve variable length as follows:
- define a tag byte which differs from all existing principal types, e.g. 255 as existing types are 1-4,
- take the longest bytes prefix of the subaccount bytes that does not end in a zero byte,
- take the byte-length of that prefix and encode it into a byte (the "length byte")
- encode the account id as:
<principal bytes> || <prefix bytes> || <length byte> || <tag>
Then it is uniquely parsable, even if the overall length is <29 bytes, and distinguishable from a pure principal.

Reply:
Regarding the encoding, bech32 is a good alternative. It looks more homogenous (no dashes), less chance to confuse with a principal. The dashes are there to aid visual comparison of two principals which probably isn't relevant for account ids.

Reply:
> We can specify in the IC interface spec that principals ending with bytes >= 128 aren't valid and will never be

That way you are occupying half of the entire space available in the type byte and taking it away from future extensions. I wouldn't do that only to save one byte in the overall length.

> For now, I still fix the length of the subaccount to be 32

I you want to do that then I would rather have you occupy two values in the type byte instead of 127. Define one for the fixed length and another one for the variable length. Especially if you want to leave the variable length for later anyway.

Reply:
> But to be honest I don't understand why we need the fixed length (32 bytes) variant. If someone uses a collison resistant hash as subaccount id then sha-224 is popular. If someone uses a pre-image resistant hash as subaccount id then 20 bytes is popular. If someone uses another principal as subaccount id then it is also shorter than 32 bytes.

Enforcing a specific length allows for more efficient storage. 32 bytes is an upper limit because some want to use SHA-256 checksum as a subaccount. Everything else you can always pad with zeros.

Reply:
> That way you are occupying half of the entire space available in the type byte and taking it away from future extensions. I wouldn't do that only to save one byte in the overall length.

Not sure we care much about the space: so far we have used only 4 tags, and one of them has never been used in practice. Even if we reserve 128 variants for variable-length, we still have 122 more tags to use. I think that's more than enough given the rate of tag usage...
But I don't have a strong opinion, <leb128 length><magic_byte> will work for me as well, it's just more work to implement.

Reply:
What I meant was that if someone has a shorter content in mind for a subaccount id, whether that's decimal 1,2,3,..., a 20 byte hash, another principal's bytes, a 28 byte hash or a 32 byte hash, you embed that into 32 bytes, i.e. pad with zeros if necessary and that is what you store. You store the full 32 bytes and that is also what the subaccount id is. The subaccount id _is_ the full 32 bytes. But now we are talking about the textual encoding of the 32 bytes long subaccount id to get a text string that you can copy-paste and send around. For that part I propose to define the encoding such that trailing zero bytes do not have to appear in the encoding. It does not affect storage but it does effect user experience. 

I think if all our account ids look like this `y7l7y-dyaaa-aaaag-qaf7a-caiaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaa`, if everyone is forced to have this length, then we are dead as a project.

Reply:
@timohanke this is just a UI representation of the full accounts. It's likely to be just copied and sent around. I don't see the problem with this representation and it has nice properties.